### PR TITLE
[Datasets] Defer first block computation when reading a Datasource with schema information in metadata

### LINKED ISSUE #33943: [Dataset] Remove pre-read computation of first block when getting schema
### What happened + What you expected to happen

In the current implementation of [ExecutionPlan._get_unified_blocks_schema](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/plan.py#L418), we force execution to compute the first block when given a `LazyBlockList`. However, when creating a Dataset from a datasource which have schema information available before reading (e.g. Parquet), this unnecessarily forces execution, since we already check for metadata in the subsequent [ensure_metadata_for_first_block](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/lazy_block_list.py#L379). Therefore, we can remove `blocks.compute_first_block()`.

### Versions / Dependencies

Ray master

### Reproduction script

```
>>> import ray
>>> ds = ray.data.read_parquet(path)
>>> print(repr(ds))
>>> ds.is_fully_executed()
True
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?

See #33943 
## Related issue number

Closes #33943 
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/plan.py [hunk 1]
```diff
@@ -412,10 +412,12 @@ def _get_unified_blocks_schema(
             fetch_if_missing: Whether to execute the blocks to fetch the schema.
         """
 
-        # Only trigger the execution of first block in case it's a lazy block list.
-        # Don't trigger full execution for a schema read.
+        # Ensure metadata is fetched for the first block. If the schema is not available
+        # in the block metadata, this will trigger the computation of the first block.
+        # Ensure the first block has schema information available in the metadata.
```
Comment (amogkam) [inline python/ray/data/_internal/plan.py:417]: comment is duplicated twice?

**Review by amogkam — APPROVED** [review]

File: python/ray/data/tests/test_dataset_parquet.py [hunk 2]
```diff
@@ -152,11 +153,11 @@ def test_parquet_read_basic(ray_start_regular_shared, fs, data_path):
         repr(ds) == "Datastream(num_blocks=2, num_rows=6, "
         "schema={one: int64, two: string})"
     ), ds
-    check_num_computed(ds, 1, 1)
+    check_num_computed(ds, 0, 0)
 
     # Forces a data read.
     values = [[s["one"], s["two"]] for s in ds.take_all()]
-    check_num_computed(ds, 2, 2)
+    check_num_computed(ds, 2, 0)
```
Comment (c21) [inline python/ray/data/tests/test_dataset_parquet.py:160]: why we have difference between bulk and streaming?

File: python/ray/data/tests/test_dataset_parquet.py [hunk 3]
```diff
@@ -475,11 +479,13 @@ def test_parquet_read_partitioned(ray_start_regular_shared, fs, data_path):
         [3, "f"],
         [3, "g"],
     ]
+    check_num_computed(ds, 2, 0)
```
Comment (c21) [inline python/ray/data/tests/test_dataset_parquet.py:482]: ditto

File: python/ray/data/tests/test_dataset_parquet.py [hunk 4]
```diff
@@ -475,11 +479,13 @@ def test_parquet_read_partitioned(ray_start_regular_shared, fs, data_path):
...
 
     # Test column selection.
     ds = ray.data.read_parquet(data_path, columns=["one"], filesystem=fs)
     values = [s["one"] for s in ds.take()]
     assert sorted(values) == [1, 1, 1, 3, 3, 3]
+    check_num_computed(ds, 2, 0)
```
Comment (c21) [inline python/ray/data/tests/test_dataset_parquet.py:488]: ditto

File: python/ray/data/tests/test_dataset_parquet.py [hunk 5]
```diff
@@ -546,11 +555,11 @@ def test_parquet_read_partitioned_explicit(ray_start_regular_shared, tmp_path):
         repr(ds) == "Datastream(num_blocks=2, num_rows=6, "
         "schema={two: string, one: int32})"
     ), ds
-    check_num_computed(ds, 1, 1)
+    check_num_computed(ds, 0, 0)
 
     # Forces a data read.
     values = [[s["one"], s["two"]] for s in ds.take()]
-    check_num_computed(ds, 2, 2)
+    check_num_computed(ds, 2, 0)
```
Comment (c21) [inline python/ray/data/tests/test_dataset_parquet.py:562]: ditto

File: python/ray/data/tests/test_dataset_parquet.py (hunk 2 above)
Comment (scottjlee) [reply python/ray/data/tests/test_dataset_parquet.py:160]: the comment here explains: https://github.com/ray-project/ray/blob/3f9b2382d158640ecb739e9bafbe01b1d6a30369/python/ray/data/tests/test_dataset_parquet.py#L37-L40

**Review by c21 — APPROVED** [review]

**Comment by c21** [comment]
Merging to master.

**Pull request merged**